### PR TITLE
[Encoder-Decoder] Force models outputs to always have batch_size as their first dim

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR remove the hard-coded variable `encoder_outputs_batch_dim_idx` from Bart and T5 by transposing BART's `encoder_outputs` dimensions before returning them.

**Reasons:**
- When adding more encoder-decoder models, we would always force the newly added model to have this variable
- When adding the modeling_encoder_decoder.py file, models that could be used in an encoder-decoder structure would also need to have this attribute, e.g. we would have to add it to Bert for example
- `encoder_outputs_batch_dim_idx` is a hard-coded variable that I don't think is very pretty

**Trade-off:**
- Now every encoder output in a encoder-decoder model has to have `batch_size` as their first dimension.

This PR is related to a question that already came up before (see #3120): 
*Should we force all model outputs to have `batch_size` as their first dimension* ?

I think it would be good to always have `batch_size` as the first dimension (to the user) exposed output @thomwolf @LysandreJik @sshleifer @julien-c

### CODE REVIEW

File: src/transformers/modeling_bart.py [hunk 1]
```diff
@@ -448,7 +450,12 @@ def forward(
 
         x = self.layernorm_embedding(x)
         x = F.dropout(x, p=self.dropout, training=self.training)
-        x = x.transpose(0, 1)  # (seq_len, BS, model_dim)
+
+        x = x.transpose(0, 1)  # (BS, seq_len, model_dim) -> (seq_len, BS, model_dim)
+        encoder_hidden_states = encoder_hidden_states.transpose(
+            0, 1
+        )  # (BS, seq_len, model_dim) -> (seq_len, BS, model_dim)
```
Comment (sshleifer) [inline src/transformers/modeling_bart.py]: (nit) can we share the comment to avoid the annoying line wrap?

**Review by sshleifer — APPROVED** [review]
LGTM!
I ran this locally -- does not change runtime for slow tests.

File: src/transformers/modeling_bart.py [hunk 2]
```diff
@@ -477,9 +483,10 @@ def forward(
             if self.output_attentions:
                 all_self_attns += (layer_self_attn,)
 
-        # Convert shapes from (seq_len, BS, model_dim) to (BS, seq_len, model_dim)
+        # Convert to standart output format: (seq_len, BS, model_dim) -> (BS, seq_len, model_dim)
```
Comment (mgoldey) [inline src/transformers/modeling_bart.py]:
```suggestion
        # Convert to standard output format: (seq_len, BS, model_dim) -> (BS, seq_len, model_dim)
```

**Review by thomwolf — APPROVED** [review]
Yes, good for me!

**Review by LysandreJik — APPROVED** [review]
LGTM!
